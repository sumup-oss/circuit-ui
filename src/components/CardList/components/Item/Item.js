/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import { isEnter, isSpacebar } from '../../../../util/key-codes';
import { sizes } from '../../../../styles/constants';

const { KILO, MEGA, GIGA } = sizes;

const sizeMap = {
  [KILO]: 'kilo',
  [MEGA]: 'mega',
  [GIGA]: 'giga'
};

const baseStyles = ({ theme }) => css`
  label: cardlist__item;

  align-items: center;
  position: relative;
  cursor: pointer;
  border-bottom: ${theme.borderWidth.kilo} solid ${theme.colors.n300};

  &:first-child {
    border-top-left-radius: ${theme.borderRadius.mega};
    border-top-right-radius: ${theme.borderRadius.mega};
  }

  &:last-child {
    border-bottom-left-radius: ${theme.borderRadius.mega};
    border-bottom-right-radius: ${theme.borderRadius.mega};
  }
`;

const getBorderStyles = theme => css`
  outline: none;

  &::after {
    content: ' ';
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    box-shadow: 0px 0px 0px ${theme.borderWidth.mega} ${theme.colors.b500};
    border-radius: ${theme.borderRadius.mega};
  }
`;

const paddingStyles = ({ theme, padding }) =>
  padding &&
  css`
    padding: ${theme.spacings[sizeMap[padding]]};
  `;

const selectedStyles = ({ theme, selected }) =>
  selected &&
  css`
    label: cardlist__item--selected;

    background: ${theme.colors.p100};

    ${getBorderStyles(theme)};
  `;

const hoverStyles = ({ theme }) => css`
  @media (hover: hover) {
    &:hover {
      ${getBorderStyles(theme)};
    }
  }

  &:focus {
    ${getBorderStyles(theme)};
  }
`;

const BaseItem = styled('div')(
  baseStyles,
  paddingStyles,
  selectedStyles,
  hoverStyles
);

const createOnKeyDown = onClick => {
  if (!onClick) {
    return null;
  }

  return event => {
    // Most clickable HTML elements can also be triggered by pressing the
    // spacebar or enter key.
    if (isEnter(event) || isSpacebar(event)) {
      onClick(event);
    }
  };
};

const Item = props => (
  <BaseItem
    aria-selected={props.selected}
    onKeyDown={createOnKeyDown(props.onClick)}
    {...props}
  />
);

Item.propTypes = {
  /**
   * When true, shows the item with selected styles.
   */
  selected: PropTypes.bool,
  /**
   * A Circuit UI spacings size.
   */
  padding: PropTypes.oneOf([KILO, MEGA, GIGA]),
  /**
   * Content of the list item.
   */
  children: PropTypes.node.isRequired
};

Item.defaultProps = {
  padding: GIGA,
  selected: false,
  tabIndex: 0
};

/**
 * @component
 */
export default Item;
