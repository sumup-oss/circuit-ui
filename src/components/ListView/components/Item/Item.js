import React from 'react';
import PropTypes from 'prop-types';
import { flow } from 'lodash/fp';

import withKeyboardEvents from '../../../../util/withKeyboardEvents';
import { sizes } from '../../../../styles/constants';
import { Hover, Wrapper } from './components';
import withAriaSelected from '../../../../util/withAriaSelected';

const { KILO, MEGA, GIGA } = sizes;

const Item = ({ children, ...props }) => (
  <Wrapper tabIndex={0} {...props}>
    {children}
    <Hover />
  </Wrapper>
);

Item.KILO = KILO;
Item.MEGA = MEGA;
Item.GIGA = GIGA;

Item.propTypes = {
  /**
   * When true, shows the item with selected styles.
   */
  selected: PropTypes.bool,
  /**
   * A Circuit UI spacings size.
   */
  padding: PropTypes.oneOf([KILO, MEGA, GIGA]),
  /**
   * Content of the list item.
   */
  children: PropTypes.node.isRequired
};

Item.defaultProps = {
  padding: Item.GIGA,
  selected: false
};

/**
 * @component
 */
export default flow(withKeyboardEvents, withAriaSelected)(Item);
