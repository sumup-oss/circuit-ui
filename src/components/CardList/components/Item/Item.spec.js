/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import Item from './Item';

describe('Item', () => {
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<Item>List item</Item>);
    expect(actual).toMatchSnapshot();
  });

  it('should render with all paddings', () => {
    ['kilo', 'mega', 'giga'].forEach(padding => {
      expect(
        create(<Item padding={padding}>List item</Item>)
      ).toMatchSnapshot();
    });
  });

  it('should render children', () => {
    const { getByTestId } = render(
      <Item>
        <span data-testid="child">Item</span>
      </Item>
    );
    const childEl = getByTestId('child');
    expect(childEl).not.toBeNull();
    expect(childEl).toHaveTextContent('Item');
  });

  it('should render with selected styles', () => {
    const actual = create(<Item isSelected>List item</Item>);
    expect(actual).toMatchSnapshot();
  });
});
