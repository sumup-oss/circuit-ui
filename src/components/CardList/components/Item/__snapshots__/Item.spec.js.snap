// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should render with all paddings 1`] = `
.circuit-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  border-bottom: 1px solid #D8DDE1;
  padding: 12px;
}

.circuit-0:first-child {
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.circuit-0:last-child {
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

@media (hover:hover) {
  .circuit-0:hover {
    outline: none;
  }

  .circuit-0:hover::after {
    content: ' ';
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    box-shadow: 0px 0px 0px 2px #3388FF;
    border-radius: 4px;
  }
}

.circuit-0:focus {
  outline: none;
}

.circuit-0:focus::after {
  content: ' ';
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  box-shadow: 0px 0px 0px 2px #3388FF;
  border-radius: 4px;
}

<div
  aria-selected="false"
  class="circuit-0 circuit-1"
  tabindex="0"
>
  List item
</div>
`;

exports[`Item should render with all paddings 2`] = `
.circuit-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  border-bottom: 1px solid #D8DDE1;
  padding: 16px;
}

.circuit-0:first-child {
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.circuit-0:last-child {
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

@media (hover:hover) {
  .circuit-0:hover {
    outline: none;
  }

  .circuit-0:hover::after {
    content: ' ';
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    box-shadow: 0px 0px 0px 2px #3388FF;
    border-radius: 4px;
  }
}

.circuit-0:focus {
  outline: none;
}

.circuit-0:focus::after {
  content: ' ';
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  box-shadow: 0px 0px 0px 2px #3388FF;
  border-radius: 4px;
}

<div
  aria-selected="false"
  class="circuit-0 circuit-1"
  tabindex="0"
>
  List item
</div>
`;

exports[`Item should render with all paddings 3`] = `
.circuit-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  border-bottom: 1px solid #D8DDE1;
  padding: 24px;
}

.circuit-0:first-child {
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.circuit-0:last-child {
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

@media (hover:hover) {
  .circuit-0:hover {
    outline: none;
  }

  .circuit-0:hover::after {
    content: ' ';
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    box-shadow: 0px 0px 0px 2px #3388FF;
    border-radius: 4px;
  }
}

.circuit-0:focus {
  outline: none;
}

.circuit-0:focus::after {
  content: ' ';
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  box-shadow: 0px 0px 0px 2px #3388FF;
  border-radius: 4px;
}

<div
  aria-selected="false"
  class="circuit-0 circuit-1"
  tabindex="0"
>
  List item
</div>
`;

exports[`Item should render with default styles 1`] = `
.circuit-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  border-bottom: 1px solid #D8DDE1;
  padding: 24px;
}

.circuit-0:first-child {
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.circuit-0:last-child {
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

@media (hover:hover) {
  .circuit-0:hover {
    outline: none;
  }

  .circuit-0:hover::after {
    content: ' ';
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    box-shadow: 0px 0px 0px 2px #3388FF;
    border-radius: 4px;
  }
}

.circuit-0:focus {
  outline: none;
}

.circuit-0:focus::after {
  content: ' ';
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  box-shadow: 0px 0px 0px 2px #3388FF;
  border-radius: 4px;
}

<div
  aria-selected="false"
  class="circuit-0 circuit-1"
  tabindex="0"
>
  List item
</div>
`;

exports[`Item should render with selected styles 1`] = `
.circuit-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  border-bottom: 1px solid #D8DDE1;
  padding: 24px;
}

.circuit-0:first-child {
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.circuit-0:last-child {
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

@media (hover:hover) {
  .circuit-0:hover {
    outline: none;
  }

  .circuit-0:hover::after {
    content: ' ';
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    box-shadow: 0px 0px 0px 2px #3388FF;
    border-radius: 4px;
  }
}

.circuit-0:focus {
  outline: none;
}

.circuit-0:focus::after {
  content: ' ';
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  box-shadow: 0px 0px 0px 2px #3388FF;
  border-radius: 4px;
}

<div
  aria-selected="false"
  class="circuit-0 circuit-1"
  tabindex="0"
>
  List item
</div>
`;
